### PR TITLE
Implemented IAM Active Validation Endpoint

### PR DESCRIPTION
**Pull request by amagana3**
This PR implements the `/rest/isIamEnabled` endpoint which is used to check whether one of the three is IAM active:

1. Machine Identity Role ARN
2. STS Credentials
3. Account / Role combination (NOT USED BY TERRAFORM PROVIDER)

I don't believe we need the third case, so I've implemented the first two cases. The first case all you need to do is pass in the `roleArn` of the machine identity. While the STS credentials can be validated by passing an empty JSON body and authenticating via STS.

If we need the last validation, let me know. Good luck! 👍🏼 

_Note: This is a WIP for this defect: [DE29276](https://rally1.rallydev.com/#/94626853604d/iterationstatus?detail=%2Fdefect%2F519106627168&fdp=true?fdp=true). I'll implement this logic in the TFP once this endpoint is released in `alks-go`._

### CODE REVIEW

File: is_iam_enabled.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package alks
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+)
+
+type IsIamRoleRequest struct {
```
Comment (webbbarker) [inline is_iam_enabled.go]: I know this is a bit of bike shedding, but I feel like these structure names do not match the their purpose very well.
```suggestion
type IsIamEnabledRequest struct {
```

File: is_iam_enabled.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	RoleArn string `json:"roleArn"`
+}
+
+// IsIamRoleResponse is used to represent a role that's IAM active or not.
+type IsIamRoleResponse struct {
```
Comment (webbbarker) [inline is_iam_enabled.go]:
```suggestion
type IsIamEnabledResponse struct {
```

File: is_iam_enabled.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+		body, err = json.Marshal(struct {
+			IsIamRoleRequest
+		}{iam})
+	} else {
+		log.Printf("[INFO] Is IAM enabled for STS: %s", c.AccountDetails.Role)
+
+		// Request for STS
+		body, err = json.Marshal(struct {}{})
+	}
```
Comment (webbbarker) [inline is_iam_enabled.go]: You can simplify this logic in the request struct by providing the JSON struct tag `omitempty`.  That tells the runtime to not serialize that type if it's a `nil` or in this case, an empty string.

File: is_iam_enabled.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+)
+
+type IsIamRoleRequest struct {
+	RoleArn string `json:"roleArn"`
```
Comment (webbbarker) [inline is_iam_enabled.go]: The comment explaining this is further down, but `omitempty` tells the serialization runtime to ignore this value if it's empty-like (empty list, `nil`, empty string, etc.).  Should simplify the code you have below to handle different call types.
```suggestion
	RoleArn string `json:"roleArn,omitempty"`
```

**Review by webbbarker — CHANGES_REQUESTED** [review]
I'd like to see the other API mode supported.  I know _technically_ this lib is primarily consumed by the ALKS TFP, but we've tried to support other use cases in the past.  Also a few comments on the changes that need review/addressing.

**Comment by amagana3** [comment]
@webbbarker Great suggestion on `omitempty`, I had no idea this is exactly what I needed and it did simplify the logic. I also appreciate you pointing out the naming - I was referencing another file and forgot to rename the structs appropriately. 🤦🏼 

For the last API mode that I didn't add logic for (Account/Role combination), the reason I didn't do so is because it requires `account` in the form of: `{{accountId}}/ALKS{{roleName}} - {{accountAlias}}` and I do not think we track the account alias. Am I mistaken here, or is there another way to retrieve that info? Maybe calling another endpoint?

**Comment by webbbarker** [comment]
If it's required, then it would just be a parameter that the library would need to take.  Looking at the ALKS code though, it appears that only the short account string is required (`account_number/role_name`).

**Comment by amagana3** [comment]
I swear I tested it and it didn't work, then I read your latest comment and tested it again and it works... :magic: I'll commit the changes here in a sec.

**Comment by amagana3** [comment]
Final update:

I implemented the logic to support account/role combination. This has now been tested in FOUR total ways:

1. Adding `roleArn` parameter while using STS creds will validate MI.
2. Omitting `roleArn` and still using STS creds will validate those creds.
3. Omitting `roleArn` and sending `account/role` combination in request body validates those creds - requires bearer token.
4. Finally, omitting `roleArn` and passing in environment variables (username, password, etc) will validate the same `account/role` combination but this time without the bearer token.

File: is_iam_enabled.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// IsIamEnabled will check if a MI, AccountDetails, or STS assumed role is IAM active or not.
+func (c *Client) IsIamEnabled(roleArn string) (*IsIamEnabledResponse, error) {
+
+	if len(roleArn) > 1 {
+		log.Printf("[INFO] Is IAM enabled for MI: %s", roleArn)
+	} else {
+		log.Printf("[INFO] Is IAM enabled for: %s", c.AccountDetails.Account)
+		log.Printf("[INFO] Is IAM enabled for: %s", c.AccountDetails.Role)
```
Comment (webbbarker) [inline is_iam_enabled.go]: This is a super nit, but let's combine this into one logger statement.
```suggestion
		log.Printf("[INFO] Is IAM enabled for: %s/%s", c.AccountDetails.Account, c.AccountDetails.Role)
```

File: is_iam_enabled.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+	iam := IsIamEnabledRequest{
+		c.AccountDetails.Account,
+		c.AccountDetails.Role,
+		roleArn,
+	}
+
+	body, err := json.Marshal(struct {
+		IsIamEnabledRequest
+	}{iam})
```
Comment (webbbarker) [inline is_iam_enabled.go]: We don't need this extra boilerplate/allocation here.  `json.Marshal` will take our request type directly.  Someday we'll get around to refactoring all the other places it's called this way, but for now let's just start doing it the proper way going forward.
```suggestion
	body, err := json.Marshal(iam)
```

File: is_iam_enabled.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+)
+
+type IsIamEnabledRequest struct {
+	Account string `json:"account,omitempty"`
+	Role    string `json:"role,omitempty"`
```
Comment (webbbarker) [inline is_iam_enabled.go]: I also missed this on my first review, but there's already an `AccountDetails` struct you can use here that has an identical format.
```suggestion
	AccountDetails
```

File: is_iam_enabled.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	RoleArn string `json:"roleArn,omitempty"`
+}
+
+// IsIamEnabledResponse is used to represent a role that's IAM active or not.
+type IsIamEnabledResponse struct {
+	BaseResponse
+	Account    string `json:"account,omitempty"`
+	Role       string `json:"role,omitempty"`
```
Comment (webbbarker) [inline is_iam_enabled.go]:
```suggestion
	Account Details
```

File: is_iam_enabled.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	if len(roleArn) > 1 {
+		log.Printf("[INFO] Is IAM enabled for MI: %s", roleArn)
+	} else {
+		log.Printf("[INFO] Is IAM enabled for: %s", c.AccountDetails.Account)
+		log.Printf("[INFO] Is IAM enabled for: %s", c.AccountDetails.Role)
+	}
+
+	iam := IsIamEnabledRequest{
+		c.AccountDetails.Account,
+		c.AccountDetails.Role,
```
Comment (webbbarker) [inline is_iam_enabled.go]: I believe this can be simplified to this, but you might need to double check.
```suggestion
		c.AccountDetails,
```

**Review by webbbarker — CHANGES_REQUESTED** [review]
Sorry Andrew.  I should've done a more through review yesterday and saved you the round-trips on this one.  I've got a couple more comments to clean up, then I think your changes are good to go.

**Comment by webbbarker** [comment]
If you want to see the `json.Marshal` stuff in action: https://play.golang.org/p/tcfIp-Pu6qs

**Comment by amagana3** [comment]
Great suggestions, especially on the `json.Marshal`. That really cleaned up the code, and I didn't know golang was smart enough to do the mapping that seamlessly. Fixed up the other suggestions too! Never a nitpick as long as it's a positive to the overall code.

**Review by webbbarker — APPROVED** [review]
I've not tested the changes, but your updates look good to me.  👍

**Pull request merged**